### PR TITLE
Add per-group-role paid-by receipt visibility filter

### PR DESCRIPTION
**Pull request by Noah231515**
## Summary

Group roles can now restrict **which receipts a member sees**, by the receipt's "paid by" user. It is opt-in and mirrors the existing category/tag grants, but is enforced at the **row level** (it hides the whole receipt rather than stripping fields off a still-visible one).

The allowed set combines a relative **"Their own receipts"** toggle (resolves to the member at query time) with zero or more specific users. Selecting only specific users (without "their own") gives a pure-reviewer role that sees someone else's receipts but not its own. An empty config is unchanged behavior — the member sees everyone's receipts.

## Backend
- New `GroupRolePaidByUserGrant` join table + `IncludeOwnPaidReceipts` column on `GroupRoleDefinition`; surfaced through `UpsertRoleCommand` / `RoleView` / swagger (desktop client regenerated).
- `GetGroupPaidByUserIdsForUser` resolves the allowed set, cached per role; the member's own id is unioned in at resolution time, never cached.
- Row-level enforcement (`services/paid_by_filter.go`): the paged list adds a per-group `WHERE` before the count — single-group `IN`, all-group disjunction so each group applies its own role — keeping `totalCount` correct; single-receipt and dependent reads deny **403** via the `HandleRequest` chokepoint; the `HasAccess` route-guard probe, search, and the group-ids list are covered too; pie chart and CSV export pass the same filter. No app-level bypass.
- Deleting a user cascades its grant rows away (the `user_id` FK is `ON DELETE CASCADE`).

## Desktop
- A "Paid-by visibility" multi-select on the group-role form (group scope only), with "Their own receipts" pinned ahead of the user list. Serializes to `includeOwnPaidReceipts` + `paidByUserGrants` and rehydrates on edit.

## Mobile
- No changes. Enforcement is server-side: mobile has no role editor, its receipts list/search are filtered automatically, and the single-receipt 403 is already handled gracefully.

## Tests
- Go unit/integration tests: resolution semantics, row-level filtering (incl. multi-group `totalCount` correctness), single-GET 403, `HasAccess` 403, and the user-delete cascade.
- Desktop role-form specs (serialize / rehydrate / reset).
- Playwright `paid-by-visibility.spec.ts`: a member restricted to their own receipts is denied (403) a hidden-payer receipt and allowed (200) their own; the list hides the hidden one and the view route redirects.

## Scope boundary
Group totals / splitting aggregates stay unfiltered (the per-viewer pie chart is filtered). This is read-visibility only — it does not restrict which payer a member may set on a receipt.

## Summary by CodeRabbit

* **New Features**
  * Added “paid-by visibility” controls for **group roles**, including a permitted-payers list and an opt-in **“their own receipts”** option.
* **Bug Fixes**
  * Enforced paid-by receipt visibility across receipt pages, lists, access checks, search, charts, and CSV exports.
  * Pagination/counts and search results now reflect only visible receipts; exports by ID are denied if any selected receipt is hidden.
  * “Amount owed” settlement totals are exempt from the paid-by filter.
* **Documentation**
  * Updated API and UI docs/OpenAPI descriptions for the new controls and semantics.
* **Tests**
  * Added unit, handler, repository, service, and E2E coverage for enforcement, edge cases, and count correctness.

### CODE REVIEW

File: api/internal/services/grant.go [hunk 1]
```diff
@@ -38,6 +38,40 @@ func (service PermissionService) GetGroupTagIdsForUser(userId uint, groupId uint
...
+// only NARROW visibility within an already-permitted group (the handler
+// permission gate is the access control). The returned map is freshly allocated
+// per call (the requesting user's id may be unioned in), so callers may keep it.
+func (service PermissionService) GetGroupPaidByUserIdsForUser(userId uint, groupId uint) (map[uint]struct{}, bool, error) {
+	entry, err := service.resolveGroupRoleGrants(userId, groupId)
+	if err != nil {
+		return nil, false, err
+	}
+	if entry == nil || (len(entry.paidByUserIds) == 0 && !entry.includeOwnPaidReceipts) {
+		return nil, true, nil
```
[the remark this answers is not in the document]
Comment (Noah231515) [reply api/internal/services/grant.go]: ACCEPT — verified and fixed in 97f0f0fa. You're right: the `user_id` FK on `GroupRolePaidByUserGrant` is `ON DELETE CASCADE`, so a role restricted to only `[X]` (include-own false) would resolve to *unrestricted* (see-all) once X was deleted and its grant row cascaded away — a silent privacy widening.

Fix: persist an explicit `PaidByVisibilityRestricted` flag on `GroupRoleDefinition`, set on save to `includeOwnPaidReceipts || len(paidByUserGrants) > 0`. `GetGroupPaidByUserIdsForUser` now keys "unrestricted" off that flag instead of the live grant count, so a configured role with no remaining payers resolves to an **empty** allowed set → the existing `IN (0)` sentinel → "see nothing" (fails closed). Added a regression test (`TestGetGroupPaidByUserIdsFailsClosedAfterGrantedUserDeleted`) that deletes the granted user and asserts the role stays restricted with an empty set.

File: api/internal/services/roles.go [hunk 2]
```diff
@@ -226,6 +235,17 @@ func validateGrantsExist(tx *gorm.DB, categoryGrants []uint, tagGrants []uint) e
...
+	if len(paidByUserGrants) > 0 {
+		userRepository := repositories.NewUserRepository(tx)
+		count, err := userRepository.CountByIds(paidByUserGrants)
+		if err != nil {
+			return err
+		}
+		if int(count) != len(paidByUserGrants) {
+			return ErrInvalidGrant
+		}
+	}
```
[the remark this answers is not in the document]
Comment (Noah231515) [reply api/internal/services/roles.go:247]: ACCEPT — fixed in 97f0f0fa. Updated both handler messages (CreateRole and UpdateRole) from "One or more category or tag grants do not exist" to "One or more category, tag, or paid-by user grants do not exist", matching `ErrInvalidGrant`'s broadened coverage.

**Comment by Noah231515** [comment]
Addressed the outside-diff finding on `api/internal/handlers/search.go` (paid-by filter applied after `Limit(100)`) in 97f0f0fa.

**ACCEPT** — verified: the query limited to 100 rows *before* the post-fetch paid-by filter, so hidden receipts filling the first 100 by date could drop a restricted user's visible matches. Fix: push the per-group paid-by predicate into the SQL query **before** the limit, via a shared `ReceiptRepository.ApplyPaidByDisjunction(query, memberGroupIds, resolver)` helper (extracted from the all-group paged read so the two paths share one implementation), and removed the now-redundant post-`Find` `FilterReceiptsByPaidBy` call for search. Added `TestApplyPaidByDisjunctionFiltersBeforeLimit` (seeds >limit hidden receipts that are newer by date + one older visible one, asserts the visible one survives a small limit).

All three review findings (this + the two inline) are addressed; full Go suite + the paid-by Playwright e2e are green.

**Comment by Noah231515** [comment]
Ran a self-review of this branch and addressed 6 findings (commit dd188c8a):

1. **Settlement stays unfiltered (correctness).** The single-receipt chokepoint was inadvertently gating `GetAmountOwedForUser` (it sets `ReceiptIds`), so a paid-by-restricted member got a 403 on amount-owed. Settlement totals must be identical for every member, so the endpoint is now exempt via a new `Handler.SkipPaidByVisibilityCheck` flag — it still enforces `group.receipts.read`, it just doesn't paid-by-filter the accounting. Browse surfaces (list/single/search/pie chart/export) stay filtered.
2. **Desktop picker reactive user pool.** The role-form's `paidByOptions` read `selectSnapshot` inside a `computed()` (non-reactive), which could drop granted users on edit if AppData landed after mount → silent grant revocation on save. Now uses `selectSignal`.
3. **One DB read instead of two** for the paid-by config flags (`GetGroupRolePaidByConfig`).
4. **enforcePaidByVisibility resolves per group**, not per receipt (reuses `FilterReceiptsByPaidBy`) — a multi-id request is now O(distinct groups).
5. **Fail closed on flag desync** — resolution honors live grant rows/include-own even if the persisted `PaidByVisibilityRestricted` flag is ever out of sync.
6. **`ApplyPaidByDisjunction` fails closed** on an empty member-group set instead of a silent no-op.

Full Go suite + the paid-by Playwright e2e (+ roles/auth smoke) are green. No API-contract/generated-client change.
